### PR TITLE
Using querySelector/querySelectorAll on elements

### PR DESCRIPTION
**Pull request by lvh**
Thanks in advance for your comments :)

### CODE REVIEW

**Comment by smashwilson** [comment]
@catlook: Care to work your magic? :sparkles:

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    - HTML
+    - JavaScript
+
+---
+
+Modern browsers have APIs called `querySelector` and `querySelectorAll`. They
+find one or more elements matching a CSS selector. I'm assuming basic
+familiarity with CSS selectors: how you select elements, classes and ids. If
+you haven't used them, the Mozilla Developer Network has an excellent
+[introduction][csssel].
```
Comment (smashwilson) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Can you add an excerpt separator somewhere, to tell Jekyll which bit to show on the index pages? Usually it's the first paragraph or so.

To do so, add an HTML comment surrounded by newlines where you want the break to be:

```

<!-- more -->

```

Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Done! Thanks :)

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+id="outside">`.
+
+With that in mind, what do these two expressions do?
+
+```javascript
+document.querySelectorAll("#my-id div div");
+document.querySelector("#my-id").querySelectorAll("div div");
+```
+
+You might reasonably expect them to be equivalent. After all, one asks
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: ... After all, one asks for ...

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+`div` elements inside `div` elements inside `#my-id`, and the other
+asks for `div` elements inside `div` elements that are *descendants*
+of `#my-id`. However, when you look at [this JSbin][jsbin] you'll see
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: ... look at [this JSbin][jsbin], you'll see ...

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+What is going on here?
+
+It turns out that [`element.querySelectorAll`][eqsa] doesn't match elements
+starting from `element`. Instead, it matches elements matching the query that
+are also descendants of `element`. Therefore, we're seeing three `div`
+elements: `div.lonely`, `div.outer`, `div.inner`. We're seeing them because
+they both match the `div div` selector and are all descendants of `#my-id`.
+
+The trick to remembering this is that CSS selectors are absolute; not relative
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: ... are absolute.  They are not relative ...

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      <div id="my-id">
+        <div class="inner"></div>
+      </div>
+    </div>
+  </body>
+</html>
+```
+
+I think this API is surprising, and the front-end engineers I've asked
+seem to agree with me. This is, however, not a bug; it's definitely
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: ... not a bug. It's definitely ...

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+how the spec claims it should work, and how it works in Firefox,
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: ... should work and how it ...

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+Chrome and Safari. [John Resig commented][jresig] how he and others
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: ... Chrome, and Safari. ...

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 2 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Fixed in ec03f20. Thanks!

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 3 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Fixed in d4e2b06. Thanks!

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 4 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Fixed in 4d1399e. Thanks!

File: _posts/2015-08-31-using-querySelector-on-elements.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+[spec]: https://dom.spec.whatwg.org/#dom-parentnode-queryselectorall
+[jresig]: http://ejohn.org/blog/thoughts-on-queryselectorall/
+[scope-spec]: https://html.spec.whatwg.org/multipage/semantics.html#attr-style-scoped
+[scope]: https://developer.mozilla.org/en-US/docs/Web/CSS/%3Ascope
+[scope-compat]: https://developer.mozilla.org/en-US/docs/Web/CSS/%3Ascope#Browser_compatibility
+[dom4-query]: http://www.w3.org/TR/dom/#interface-parentnode
+[older-dom4]: http://www.w3.org/TR/2014/WD-dom-20140204/#interface-parentnode
+[dom4-polyfill]: https://webreflection.github.io/dom4/
+[dom-elements-polyfill]: https://github.com/barberboy/dom-elements
+[living-dom]: https://dom.spec.whatwg.org/#interface-parentnode
```
Comment (catlook) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: [ Ed note:  Cool!  Use the edits you like and discard the others.  Let me know if you have questions!  - Cat ]

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 5 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Fixed in f2c4ca5. Thanks!

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 6 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: I got rid of this sentence in 7cd0ab1.

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 7 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: I got rid of this sentence in 7cd0ab1.

**Comment by smashwilson** [comment]
Boom. :shipit:

File: _posts/2015-08-31-using-querySelector-on-elements.md (hunk 8 above)
Comment (lvh) [inline _posts/2015-08-31-using-querySelector-on-elements.md]: Thanks a lot! I really appreciate the help of an experienced editor :) I know I use semicolons a little too much. The versions without definitely read better. Without the Oxford comma, I keep tripping over the sentence about browsers implementing the spec, but that phrase was kind of superfluous anyway, so I got rid of it.

**Comment by lvh** [comment]
Awesome, thanks y'all!

**Comment by smashwilson** [comment]
@lvh: Thanks for contributing! It'll go live as soon as the build finishes on master (which can take a bit).

**Comment by rgbkrk** [comment]
Oh yes, living in the future. I :heart: me some `document.querySelector`.

**Pull request merged**